tests/test_composite_logprob.py

```diff
@@ -0,0 +1,97 @@
+
+
+def test_scalar_clipped_mixture():
+    x = at.clip(at.random.normal(loc=1), 0.5, 1.5)
+    x.name = "x"
+    y = at.random.beta(1, 2, name="y")
+
+    comps = at.stack(x, y)
+    comps.name = "comps"
+    idxs = at.random.bernoulli(0.4, name="idxs")
+    mix = comps[idxs]
+    mix.name = "mix"
+
+    mix_vv = mix.clone()
+    idxs_vv = idxs.clone()
+    logp = joint_logprob({idxs: idxs_vv, mix: mix_vv})
+
+    logp_fn = aesara.function([idxs_vv, mix_vv], logp)
+    assert logp_fn(0, 0.4) == -np.inf
+    assert np.isclose(logp_fn(0, 0.5), st.norm.logcdf(0.5, 1) + np.log(0.6))
+    assert np.isclose(logp_fn(0, 1.3), st.norm.logpdf(1.3, 1) + np.log(0.6))
+    assert np.isclose(logp_fn(1, 0.4), st.beta.logpdf(0.4, 1, 2) + np.log(0.4))
+
+
+def test_nested_scalar_mixtures():
+    x = at.random.normal(loc=-50, name="x")
+    y = at.random.normal(loc=50, name="y")
+    comps1 = at.stack(x, y)
+    comps1.name = "comps1"
+    idxs1 = at.random.bernoulli(0.5, name="idxs1")
+    mix1 = comps1[idxs1]
+    mix1.name = "mix1"
+
+    w = at.random.normal(loc=-100, name="w")
+    z = at.random.normal(loc=100, name="z")
+    comps2 = at.stack(w, z)
+    comps2.name = "comps2"
+    idxs2 = at.random.bernoulli(0.5, name="idxs2")
+    mix2 = comps2[idxs2]
+    mix2.name = "mix2"
+
+    comps12 = at.stack(mix1, mix2)
+    comps12.name = "comps12"
+    idxs12 = at.random.bernoulli(0.5, name="idxs12")
+    mix12 = comps12[idxs12]
+    mix12.name = "mix12"
+
+    idxs1_vv = idxs1.clone()
+    idxs2_vv = idxs2.clone()
+    idxs12_vv = idxs12.clone()
+    mix12_vv = mix12.clone()
+
+    logp = joint_logprob(
+        {idxs1: idxs1_vv, idxs2: idxs2_vv, idxs12: idxs12_vv, mix12: mix12_vv}
+    )
+    logp_fn = aesara.function([idxs1_vv, idxs2_vv, idxs12_vv, mix12_vv], logp)
+
+    expected_mu_logpdf = st.norm.logpdf(0) + np.log(0.5) * 3
+    assert np.isclose(logp_fn(0, 0, 0, -50), expected_mu_logpdf)
+    assert np.isclose(logp_fn(0, 1, 0, -50), expected_mu_logpdf)
+    assert np.isclose(logp_fn(1, 0, 0, 50), expected_mu_logpdf)
+    assert np.isclose(logp_fn(1, 1, 0, 50), expected_mu_logpdf)
+    assert np.isclose(logp_fn(0, 0, 1, -100), expected_mu_logpdf)
+    assert np.isclose(logp_fn(0, 1, 1, 100), expected_mu_logpdf)
+    assert np.isclose(logp_fn(1, 0, 1, -100), expected_mu_logpdf)
+    assert np.isclose(logp_fn(1, 1, 1, 100), expected_mu_logpdf)
+
+    assert np.isclose(logp_fn(0, 0, 0, 50), st.norm.logpdf(100) + np.log(0.5) * 3)
+    assert np.isclose(logp_fn(0, 0, 1, 50), st.norm.logpdf(150) + np.log(0.5) * 3)
+
+
+@pytest.mark.xfail(
+    reason="Measurable rewrites are too eager, when we have a RV in the graph that should not be measured"
+)
+def test_overeager_rewrites():
+    x = at.random.normal()
+    x.tag.ignore_logprob = True
+    y = at.clip(x, 0, 1)
+    z_rv = at.random.normal(y, 1, name="z")
+    z_vv = z_rv.clone()
+
+    logp = joint_logprob({z_rv: z_vv})
+
+    assert not any(
+        isinstance(node.owner.op, CensoredRV)
```

Comment:
This is not an issue in this case as the `CensoredRV` `Op` is just a subclass of the original `Clip` `Op`, but this may not always be the case for our IR, and could become unnecessarily costly as we add more support for basic `Op`s to the limit we might be recreating the whole graph of deterministics with "measurable" equivalents.

Reply:
This is the issue I was referring to (indirectly) [here](https://matrix.to/#/!RcUetDjlPGYsTqnQDE:gitter.im/$Ct4DgOBWJHqo7kVWpI2X0JCSHO7hiKVv8Ll_FbZ0BRU?via=gitter.im&via=matrix.org).

My first thought was that we could remove/undo unvalued measurable transforms during some "post-processing" pass of the IR rewrites, or during/after construction of the log-probability graphs.  The latter sounds better, because then we would have much simpler "measurable" IR graphs; ones that could be used for other things that require these "non-computable" measurable `Op`s.

Your comment raises another important point: we should try to standardize the measurable `Op`s&mdash;even the ones that are simple views/clones/wrappers of other `Op`s.  One of the constraints is that such measurable `Op`s essentially act like views/clones of the original `Op`s/`Apply` nodes but without introducing unnecessary indirection (e.g. like using the actual `View` `Op` would).  The subclassing approach seems to work well enough, so we can probably stick with that; however, we need a clean, unified interface for it.

At least one part of this standardization could be addressed through extensions of the `aeppl.abstract.assign_custom_measurable_outputs` constructor function approach.  For instance, we could create a `Variable`-based interface for that constructor that handles all [these steps](https://github.com/aesara-devs/aeppl/blob/a97a6bcbd60726bb5142b39c4da52c7990d50311/aeppl/mixture.py#L281-L283) and only use that constructor wherever we need to convert an existing `Op` to its measurable IR.

Reply:
Yes I think we can easily automate most of those steps. In many cases the original Op / node have all the info needed for the logprob function. Some cases that need special inputs like Mixtures perhaps (which can have different original forms), can have custom defined classes as IR but I think those will be the exception rather than the norm.